### ISSUE TITLE
Allow CSS selectors to return the full html content, not just the text within it.

### ISSUE THREAD

**Issue by 0xdevalias** [sibprogrammer/xq#20]
The ability to use CSS Selectors were added in:

- https://github.com/sibprogrammer/xq/issues/16

Which references the following 2 commits:

- https://github.com/sibprogrammer/xq/commit/8dbf6ee9e9fc9d589cf5cfe0a8cc23038706d014
- https://github.com/sibprogrammer/xq/commit/35ea9e317ea07bbeed70d97b36b17d31349d3214

The docs show that we can extract content using CSS selectors as follows:

```
xq -q "body > p"
```

Yet this seems to return the `.innerText` equivalent, not the actual HTML structure we're querying.

Looking at the code in `internal/utils/utils.go` it looks like [`goquery`](https://github.com/PuerkitoBio/goquery) is being used (`goquery` internally makes use of [`cascadia`](https://github.com/andybalholm/cascadia) for it's CSS selector support):

https://github.com/sibprogrammer/xq/blob/8dbf6ee9e9fc9d589cf5cfe0a8cc23038706d014/internal/utils/utils.go#L173-L184

Specifically it seems to be using `.Text()` to return the `.innerText` equivalent of the selected node:

- https://pkg.go.dev/github.com/PuerkitoBio/goquery#Selection.Text
  - > Text gets the combined text contents of each element in the set of matched elements, including their descendants.

Yet looking at the `goquery` docs, it's possible to access the `.innerHTML` (or even 'outer HTML') for a selected node:

- https://pkg.go.dev/github.com/PuerkitoBio/goquery#Selection.Html
  - > Html gets the HTML contents of the first element in the set of matched elements. It includes text and comment nodes.
- https://pkg.go.dev/github.com/PuerkitoBio/goquery#OuterHtml
  - > OuterHtml returns the outer HTML rendering of the first item in the selection - that is, the HTML including the first element's tag and attributes.
  - > Unlike InnerHtml, this is a function and not a method on the Selection, because this is not a jQuery method (in javascript-land, this is a property provided by the DOM).
  
---
  
So basically, to implement this feature, it looks like we just need to use the `.Html()` function instead of the `.Text()` function; most likely also adding a CLI argument that controls when one is used over the other.

**Comment by 0xdevalias** [comment]
As a workaround in the meantime, it looks like the [`cascadia`](https://github.com/andybalholm/cascadia) library has a separate CLI tool:

- https://github.com/andybalholm/cascadia
  - > CSS selector library in Go
- https://github.com/suntong/cascadia
  - > The [Go Cascadia package](https://github.com/andybalholm/cascadia) implements CSS selectors for html. This is the command line tool, started as a thin wrapper around that package, but growing into a better tool to test CSS selectors without writing Go code:

**Comment by madneal** [comment]
Thanks for your suggestions, but cli tool will not work in chrome extension

**Comment by 0xdevalias** [comment]
> Thanks for your suggestions, but cli tool will not work in chrome extension

@madneal I'm aware of that. Using the CLI is just my workaround to your extension currently being broken, as I described here:

- https://github.com/madneal/export-to-markdown/issues/14#issuecomment-1374702021

For actually fixing that issue within your extension you don't need this or any other CLI tools/library/etc because the required functionality is natively built into the browser, and thus usable by your extension.

**Comment by 0xdevalias** [comment]
Just found my way back to this issue again after trying to use `xq` in this way, and wondering why it wasn't outputting anything from my `--query`.

Back to the workaround it is then ([Ref](https://github.com/sibprogrammer/xq/issues/20#issuecomment-1374673443))

**Comment by sibprogrammer** [comment]
@0xdevalias could you provide a little bit more detailed example?
If we are talking about example.com there we should use some different selector instead of "body > p". E.g.
`curl -s https://example.com/ | ./xq -q 'body > div > p'` or `curl -s https://example.com/ | ./xq -q 'p'`

**Comment by 0xdevalias** [comment]
@sibprogrammer Of course :)

You can check out the comments I made over at the `cascadia` repo (Ref: [1](https://github.com/suntong/cascadia/issues/3#issuecomment-1590736749), [2](https://github.com/suntong/cascadia/issues/3#issuecomment-1593927751), etc) (as it turns out it didn't solve my needs either this time due to some bugs/limitations, so I ended up using `pup` (Ref: [1](https://github.com/suntong/cascadia/issues/3#issuecomment-922444518), [2](https://github.com/ericchiang/pup)) instead)

This time around, basically I wanted to process something like this:

```html
<html>
<head>
  <script src="foo.js"></script>
  <script src="bar.js"></script>
  <script src="baz.js"></script>
</head>
</html>
```

By using a query something like:

```css
html > head > script[src]
```

Be able to get a response back like this:

```
<script src="foo.js"></script>
<script src="bar.js"></script>
<script src="baz.js"></script>
```

And then ideally, if I was also able to then tell it to return the contents of the attribute `src`, my end goal as to get output like this:

```
foo.js
bar.js
baz.js
```

---

I proposed what I thought would be useful syntax/semantics for this over on `cascadia` (Ref: [1](https://github.com/suntong/cascadia/issues/3#issuecomment-1594038735, [2](https://github.com/suntong/cascadia/issues/3#issuecomment-1600034259)); which already has a concept of "output this sub-query" (using `--piece`), but essentially it would be by being able to use the same `[foo]` CSS attribute selector syntax in a different context (in their case, within `--piece` as opposed to `--query`) to say "output the `foo` attribute from all matching elements that have it)

- https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors

> > I didn't look into the url closely, but having check it out again just now, I found that the `a[title]` in CSS Attribute selectors means _`<a>` elements **with** a title attribute_, IE, how to select the `<a>` elements, while here we need a syntax to return attributes, which is not supported by CSS selectors (still).
> 
> Yup, that is how it works in normal CSS selector usage, and is how it would need to work (and just how it does work I believe) in the `--query` part of `cascadia`; but what I was proposing above is that we could leverage the same familiar semantics of that syntax, but when used within `--piece`, we could have it output the actual attribute being described.
> 
> So basically:
> 
> - `--query`:
>   - `[foo]` would return all elements that have an attribute named `foo`
>   - `a[foo]` would return all elements that have an `a` attribute named `foo`
>   - etc
> - `--piece`
>   - `[foo]` would return the attribute named `foo` from all elements that have it
>   - `a[foo]` would return the attribute named `foo` from all `a` elements that have it
>   - etc
> 
> So then combined, you could use these like I described above:
> 
> > That way the `--piece` syntax would sort of end up being much closer to just standard CSS selector usage. And I could do something like:
> > 
> > ```shell
> > ⇒ pbpaste | cascadia --in --out --css 'html > head > script' --piece url='[src]'
> > 
> > ⇒ pbpaste | cascadia --in --out --css 'html > head' --piece url='script[src]'
> > 
> > # etc
> > ```
> > 
> > _Originally posted by @0xdevalias in https://github.com/suntong/cascadia/issues/3#issuecomment-1594038735_
> 
> _Originally posted by @0xdevalias in https://github.com/suntong/cascadia/issues/3#issuecomment-1600034259_

There is also the 'prior art' of how `pup` has chosen to implement their 'Display Functions' (as they call them):

- https://github.com/ericchiang/pup#display-functions

Which uses `attr{attrkey}` (eg. `attr{foo}`):

- https://github.com/ericchiang/pup#attrattrkey
  - > Print the values of all attributes with a given key from all selected nodes.

**Comment by sibprogrammer** [comment]
@0xdevalias thank you for clarification!
I prototyped a straightforward solution (didn't publish it yet):
```
*[master][~/projects/xq]$ cat ~/tmp/text.html | ./xq -q "head > script[src]" -a "src"
foo.js
bar.js
baz.js
*[master][~/projects/xq]$ cat ~/tmp/text.html | ./xq -q "head > script" -a "src"
foo.js
bar.js
baz.js
```
It seems it should be enough for attribute extraction. pup syntax looks more promising, but more complicated (at least at first glance).

**Comment by 0xdevalias** [comment]
> I prototyped a straightforward solution (didn't publish it yet):

@sibprogrammer Nice, looks good! Will there also be a longer `--attr` or similar as well as the `-a`?

> It seems it should be enough for attribute extraction. pup syntax looks more promising, but more complicated (at least at first glance).

_nods_ yeah, that's fair. Personally I don't know if a lot of the `pup` syntax would even matter / be relevant to my day to day needs anyway. Just thought I would include it in case it was useful.
